src/bundler/transformer.ts

```diff
@@ -83,12 +83,13 @@ export class Transformer {
         };
         async.eachSeries(transforms, (transform: Transform, onTransformApplied: ErrorCallback<Error>) => {
             process.nextTick(() => {
-                transform(context, (error: Error, dirty: boolean) => {
+                transform(context, (error: Error, dirty: boolean, transpile: boolean = false) => {
```

Comment:
We need a new parameter here, `transpile` indicates whether the source should be transpiled again or not, we can't have a parameter that sometimes changes what it means. https://github.com/jods4/karma-typescript/blob/master/src/api/transforms.ts#L7

This new parameter (`isScript` or similar) could be used together with the `transpile` parameter to check if an `ast` should be created or not.

Reply:
Sorry I'm not sure I understood what you mean? Do you want a fourth parameter?

Reply:
Yeah, a fourth parameter: the api is shared between Typescript and Javascript transforms and the `transpile` parameter is used to determine whether the source should be transpiled again or not.

Maybe it would be better to use just three parameters and add two new interfaces to the api:

```js
export interface JsTransformCallback {
    (error: Error, dirty: boolean, transformedScript?: boolean): void;
};

export interface JsTransform {
    (context: TransformContext, callback: JsTransformCallback): void;
    initialize?: TransformInitialize;
};
```

And then rename the third parameter in the js transform callback to `transformedScript: boolean = false`.

What I'm looking for here is keeping the transforms api as simple an uncluttered as possible while keeping the ts and js transforms as separated as possible, it's a bit of a balancing act, I know ☺️

The js transform would look something like this:
```js
async.eachSeries(transforms, (transform: JsTransform, onTransformApplied: ErrorCallback<Error>) => {
    process.nextTick(() => {
        transform(context, (error: Error, dirty: boolean, transformedScript: boolean = false) => {
            this.handleError(error, transform, context);
            if (dirty) {
                bundleItem.ast = context.js.ast;
                bundleItem.source = context.source;
            }
            bundleItem.transformedScript = transformedScript;
            onTransformApplied();
        });
    });
}, onTransformsApplied);
```

I'll document this when we're done 🚀

Reply:
I tried that but it doesn't work well because then you have `TsTransform` and `JsTransform` that are different, yet they "unify" in `configuration.bundler.transforms`.
I'd say it's best to somehow go with your first idea and keep the transforms identical.